server/src/server/mod.rs

```diff
@@ -60,6 +61,7 @@ impl Server {
         let is_graceful_shutdown = Arc::new(AtomicBool::new(false));
 
         let on_start_fut = on_worker_start();
+        let on_stop_fut = on_worker_stop();
```

Comment:
the closure executes eagerly on worker start meaning 
```rust
|| {
// happens when worker start.
async {
    // happens when worker stop.
}
}
``` 
This can be a potential footgun.

Reply:
Yup it shot me as well. Should we make this more similar to tokio's on_thread_start/on_thread_stop callbacks that just fake a non-async function?

Reply:
My opinion is we make it accepting a future rather than a closure. Reason being worker itself runs in an async context already so there is no need to do `async {  closure().await }` and we can just do `async { future.await }`.
`on_worker_start` and tokio's case are different because these APIs have access to non async context (or partially so )

Reply:
Unfortunately I can't think of how that would work since the future would need to be awaited multiple times, once per worker task thread and that's not a power that future's have. I think that's why you used a generator pattern where the callback vends a new future each time and why Tokio doesn't have this problem since a normal Fn can be invoked multiple times.

Reply:
If you don't mind waiting on this PR for a bit we can switch to async closure when Rust 1.85 release later this month. xitca as a whole would switch to Rust 1.85 and edition 2024 at the same time. After which we can use `on_worker_stop(async || {})` so there is no accidental eager execution on the closure.